### DISCUSSION TITLE
I don't know what is option's argument

### DISCUSSION THREAD

**Discussion by blusewill** [openai/whisper#1229] [Q&A] [answered]
So I am writing a python script to generate SRT file
and I don't know how it's won't available to generate the file
It's will keep saying **TypeError: __call__() missing 1 required positional argument: 'options'** and I checked every single forum thread but there is no similar one that I encountered, hope someone can help me to fix this issue

![](https://imgur.com/xwiiXCv.png)

**Comment by phineas-pta** [comment]
this is a new feature in development, you should instead use the stable version, in colab, rather than

```
pip install git+https://github.com/openai/whisper.git
```

you do

```
pip install openai-whisper
```

**Comment by g-l-i-t-c-h-o-r-s-e** [comment]
> this is a new feature in development, you should instead use the stable version, in colab, rather than
> 
> ```
> pip install git+https://github.com/openai/whisper.git
> ```
> 
> you do
> 
> ```
> pip install openai-whisper
> ```

What is the `options` argument though?

**Reply by erkkimon** [reply]
I'm wondering the same. It would be nice to know how to use the latest version, because I bet it's not really broken.

**Reply by erkkimon** [reply]
After taking a further look, it really seems broken. Looks like the options will do stuff in the future but not yet, so many functions expect options but internally it isn't passed, so you can't pass them to the expecting functions without modifying the source code.

Here's an ugly fix which works for JSON format. You can delete options for other writing functions as well if you wish. This works if you call get_writer like this this:

`writer = get_writer("json", "./output")`

And here comes the ugly fix in case someone needs.

```
--- utils.py.original	2023-09-21 15:29:45.437190286 +0300
+++ utils.py.modified	2023-09-21 15:28:54.433188605 +0300
@@ -74,7 +74,7 @@
     def __init__(self, output_dir: str):
         self.output_dir = output_dir
 
-    def __call__(self, result: dict, audio_path: str, options: dict):
+    def __call__(self, result: dict, audio_path: str):
         audio_basename = os.path.basename(audio_path)
         audio_basename = os.path.splitext(audio_basename)[0]
         output_path = os.path.join(
@@ -82,9 +82,9 @@
         )
 
         with open(output_path, "w", encoding="utf-8") as f:
-            self.write_result(result, file=f, options=options)
+            self.write_result(result, file=f)
 
-    def write_result(self, result: dict, file: TextIO, options: dict):
+    def write_result(self, result: dict, file: TextIO):
         raise NotImplementedError
 
 
@@ -231,7 +231,7 @@
 class WriteJSON(ResultWriter):
     extension: str = "json"
 
-    def write_result(self, result: dict, file: TextIO, options: dict):
+    def write_result(self, result: dict, file: TextIO):
         json.dump(result, file)
 
 
@@ -249,9 +249,9 @@
     if output_format == "all":
         all_writers = [writer(output_dir) for writer in writers.values()]
 
-        def write_all(result: dict, file: TextIO, options: dict):
+        def write_all(result: dict, file: TextIO):
             for writer in all_writers:
-                writer(result, file, options
)
+                writer(result, file, {})
 
         return write_all
```

PS. I added a modified utils.py as attachments and a working example of working call as attachments.

[utils.py.modified.txt](https://github.com/openai/whisper/files/12684581/utils.py.modified.txt)
[test.py.txt](https://github.com/openai/whisper/files/12684590/test.py.txt)

**Comment by bradsec** [comment, marked as answer]
Had the same issue trying to integrate into a python program and save different output format types `.srt` `.json` etc - even using `pip install git+https://github.com/openai/whisper.git --upgrade`. Running as a CLI seemed to work fine, however trying to run like in the below code, regardless of file type  the options error always displayed `TypeError: ResultWriter.__call__() missing 1 required positional argument: 'options'`

```
def whisper_transcription(input_audio, whisper_model):

    model = whisper.load_model(whisper_model)
    result = model.transcribe(input_audio, verbose=True)
    output_directory = "."

    txt_writer = get_writer("txt", output_directory)
    txt_writer(result, input_audio)

    json_writer = get_writer("json",  output_directory)
    json_writer(result, input_audio)

    vtt_writer = get_writer("vtt", output_directory)
    vtt_writer(result, input_audio)

    tsv_writer = get_writer("tsv", output_directory)
    tsv_writer(result, input_audio)

    srt_writer = get_writer("srt", output_directory)
    srt_writer(result, input_audio)
```

Error seemed to be caused by this section of the Whisper `utils.py` when no values or default values were being received by some of the options - 
```
class SubtitlesWriter(ResultWriter):
    always_include_hours: bool
    decimal_marker: str

    def iterate_result(self, result: dict, options: dict):

        raw_max_line_width: Optional[int] = options["max_line_width"] # This line
        max_line_count: Optional[int] = options["max_line_count"] # This line
        highlight_words: bool = options["highlight_words"] # This line
        max_line_width = 1000 if raw_max_line_width is None else raw_max_line_width
        preserve_segments = max_line_count is None or raw_max_line_width is None
```

As a workaround I made the following changes to my application, and passed in some default options values. 

```
def whisper_transcription(input_audio, whisper_model):

    model = whisper.load_model(whisper_model)
    result = model.transcribe(input_audio, verbose=True)
    output_directory = "."

    # Set some initial options values
    options = {
        'max_line_width': None,
        'max_line_count': None,
        'highlight_words': False
    }

   # Include options in call
    txt_writer = get_writer("txt", output_directory)
    txt_writer(result, input_audio, options)

    json_writer = get_writer("json",  output_directory)
    json_writer(result, input_audio, options)

    vtt_writer = get_writer("vtt", output_directory)
    vtt_writer(result, input_audio, options)

    tsv_writer = get_writer("tsv", output_directory)
    tsv_writer(result, input_audio, options)

    srt_writer = get_writer("srt", output_directory)
    srt_writer(result, input_audio, options)
```

Another option may just be to construct args in python and use the CLI whisper command formatting.

**Comment by blusewill** [comment]
Never Thought I am going to open this discussion again.
Because it happen again **NOT ON GITHUB RELEASE** but actually stable release
![image](https://github.com/openai/whisper/assets/62820836/d5ccaf5f-eadd-4129-aa8b-ace546e67127)

The Code can be found at Here https://github.com/blusewill/ytvideo-whisper/tree/dev

**Comment by glangford** [comment]
@blusewill There was a breaking change made to this API. @Mdsajal5 and @bradsec already documented workarounds above so this is just another way of solving the problem.

Depending on your need you can either call
```python
writer( result, file, {} )
```

or with the complete set of options, for SRT files for example:
```python
writer(result, file, {"max_line_width":55, "max_line_count":2, "highlight_words":False} ) 
```

The `options` dictionary is checked in the first few lines of this code in whisper, fyi.
https://github.com/openai/whisper/blob/0a60fcaa9b86748389a656aa013c416030287d47/whisper/utils.py#L103

**Comment by BigApe99** [comment]
this error went away when I installed a certain version of whisper:
`pip install openai-whisper==20230314`

**Comment by blusewill** [comment]
@glangford Thanks for Mentioning. I am currently testing these options and see it will work in my code or not

**Comment by mlane3** [comment]
these directions are not clear.  I am going to make a new support ticket as a bug.
